Add: indicate locked rooms in exits dialogue

SlySven opened the pull request:
This PR was prompted by a discussion in Discord about the locked status of rooms not being shown in the "Room exits dialogue" despite it having a fundamental effect on the usability of exits in route-finding/speed-walking:

See Kaylee#0064's comment in the "#help" channel on 2022/05/05: https://www.discord.com/channels/283581582550237184/283582068334526464/971547534994591794 and the subsequent couple of comments from @vadi2 and then me on it.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Add: indicate locked rooms in exits dialogue

This PR was prompted by a discussion in Discord about the locked status of
rooms not being shown in the "Room exits dialogue" despite it having a
fundamental effect on the usability of exits in route-finding/speed-walking:

See Kaylee#0064's comment in the "#help" channel on 2022/05/05:
https://www.discord.com/channels/283581582550237184/283582068334526464/971547534994591794
and the subsequent couple of comments from Vadim and then me on it.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
/refresh links

SlySven commented:
/refresh links

vadi2 commented on src/dlgRoomExits.cpp:
```diff
@@ -1481,6 +1521,15 @@ void dlgRoomExits::init()
+        // status of the room:
+        roomID->addAction(mpAction_exitRoomLocked, QLineEdit::TrailingPosition);
+        // Also revise the tool tip:
+        roomID->setToolTip(utils::richText(tr("This is the Room ID number for this room; this <b>room is locked</b> so it will not be used for speed-walks at all. Neither of these details can be changed here!",
```
```suggestion
        roomID->setToolTip(utils::richText(tr("This is the Room ID number for this room; this <b>room is locked</b> so it will not be used for speed-walks at all.",
```

SlySven replied:
👎 It isn't clear from the `QLineEdit` widget used that it is set to read-only mode - this text (and it's predecessor in the prior code), and the "Forbidden" cursor, is used to give the end-user a clue why they cannot change the room's number.

vadi2 replied:
I don't think this is still helpful. Let's wait for feedback from the field - eg people being confused about this - before adding it. The fact that it's read only is enough of a hint, a forbidden cursor is going overboard and a tooltip on top of that is just not aesthetic anymore imho.

SlySven replied:
A `QLineEdit` that is set to **read-only** mode is NOT visibly different from a normal one that can be written in. (This is NOT the same as the appearance when it is *disabled*!) This is why we have some horrible, incompatible with style-sheets, messing about with `QPalette`s to change the colouration for some of the `QLineEdit`s in the "Connect profile" dialogue BTW. Therefore, providing hints, in the form of the "forbidden" cursor and explaining that the entry cannot be changed - ***has been present on that field for many years already*** and is NOT being added here!

vadi2 replied:
What do others think? @demonnic @keneanung

demonnic replied:
Is there something other than a `QLineEdit` widget that can be used to display this in a way that is obviously not editable? That seems to me like it'd be the best solution. Then the tooltip would definitely not be needed. I admit I'm not super familiar with the widgets available though.

SlySven replied:
:thinking: I suppose we could just use `QLabel`s for those - either just by extending the title labels (the red things below) to the left of the `QLineEdit`s) to display the value as well as the title - or adding additional ones to display the values only - the latter will probably be simpler to do to maintain the alignments (right for the titles, left for the values):

![Screenshot_20220616_233110](https://user-images.githubusercontent.com/6163092/174189588-650559f6-f49a-4d2e-898a-318f2f9b9b24.png)

vadi2 replied:
Seems like something that could work - could you push the change so we can preview it live?

SlySven replied:
Done in 3f8b8c1 .

vadi2 replied:
Looks fine, but the tooltip is still a bit intense, no?

![image](https://user-images.githubusercontent.com/110988/174425100-70875940-7fc7-4ba4-bcda-ba5ca9a7fcbf.png)

SlySven replied:
I think not, as it happens it is ***not*** possible to change the Room ID *anywhere* but given how many other things **can** be modified through this dialogue I do not believe it to be *that* excessive...

BTW Have you inspected how things look when the room is locked? The tool-tip is modified in that case to explain the significance...

vadi2 replied:
Yes, - 
![image](https://user-images.githubusercontent.com/110988/174474725-081b61ec-3dcb-40ed-860d-f162abc12507.png)

vadi2 replied:
I still think we should apply the original suggestion - it's more concise.

vadi2 commented on src/dlgRoomExits.cpp:
```diff
@@ -1044,9 +1044,24 @@ QAction* dlgRoomExits::getActionOnExit(QLineEdit* pExitLineEdit) const
     if (exitRoomName.trimmed().length()) {
+        if (exitRoomLocked) {
+            if (outOfAreaExit) {
+                return doubleParagraph.arg(tr("Exit to \"%1\" in area: \"%2\".")
```
Can you say what kind of an exit it is? For examle "East exit to ..."

This will help clarity in the dialog to know what kind of an exit it is, since you can't immediately tell this from the dialog, for example we have no idea where any of these rooms are going:

![image](https://user-images.githubusercontent.com/110988/167785563-4b7b8143-a63a-4245-b382-5acaf9c78d36.png)

SlySven replied:
Don't forget "tool-tip-text" generator is used for all the normal and the special exit roomID number display widgets!

It would be tricky (I think) to include the directional detail in the tool-tip and yet avoid assembling sentences for what will, don't forget, be translated, from "fragments" of text. It would be simpler if an indication of the direction, beyond the place-holder texts on non-existent exits, is really wanted to change the `QFrame`s around the controls for each normal exit direction to `QGroupBox`es (without the `QCheckBox` detail) but that would make the overall dialogue (a bit) larger...

> ... for example we have no idea where any of these rooms are going:

Other than their relative placement in the dialogue!

vadi2 replied:
Yeah, and the relative placement is not the most intuitive thing that jumps out at you. Without context here, we really have no idea!

SlySven replied:
I've looked and because i refactored the code a while back to use the same tool-tip text generator for all the places where a destination room id is entered, means it is not simple to introduce a *translatable* text fragment into the tool tips for normal exit direction.

I've looked and TBH it is much simpler to just switch to `QGroupBox`s - so that is what I am doing...

SlySven replied:
... in the separate PR #6104.

vadi2 commented on src/dlgRoomExits.cpp:
```diff
@@ -421,28 +437,6 @@ void dlgRoomExits::slot_editSpecialExit(QTreeWidgetItem* pI, int column)
 void dlgRoomExits::slot_addSpecialExit()
 {
     auto pI = new QTreeWidgetItem(specialExits);
-// MINDE
```
👍

vadi2 commented on src/dlgRoomExits.cpp:
```diff
@@ -965,15 +959,25 @@ void dlgRoomExits::setIconAndToolTipsOnSpecialExit(QTreeWidgetItem* pSpecialExit
-        pSpecialExit->setIcon(ExitsTreeWidget::colIndex_exitStatus, showIconOnExitStatus ? (outOfAreaExit ? mIcon_otherAreaExit : mIcon_inAreaExit) : QIcon());
+        // status icons)
+        const QString roomIdToolTipText{generateToolTip(pExitToRoom->name, exitAreaName, exitRoomLocked, outOfAreaExit, pExitToRoom->getWeight())};
+        pSpecialExit->setToolTip(ExitsTreeWidget::colIndex_exitStatus, utils::richText(tr("A positive roomID of the room that this special exit leads to is expected here. "
```
We don't really have to worry about positive/negative roomID's in Mudlet, it's not an issue folk struggle with, so we can shorten this -

```suggestion
        pSpecialExit->setToolTip(ExitsTreeWidget::colIndex_exitStatus, utils::richText(tr("roomID of the room that this special exit leads to is expected here. "
```

SlySven replied:
Since it is an instruction giving guidance to the end-user it make sense to say what WILL work here, it is perfectly possible to enter a negative number (and indeed, slipping a `-` sign in might be the quickest way to invalidate an existing entry...); also, starting the sentence with "roomID" doesn't look right to me, as an English speaker I'd expect it to be capitalised - with the "A positive ..." before it though I avoided having to chose between "roomID" or "RoomID" 😜

vadi2 replied:
We want to be as concise as possible in these things, not verbose.

SlySven replied:
Grammatically, if we are losing the "positive" then we still need the definite article preceding `roomID`:
```suggestion
        pSpecialExit->setToolTip(ExitsTreeWidget::colIndex_exitStatus, utils::richText(tr("The roomID of the room that this special exit leads to is expected here. "
```

SlySven replied:
Done in https://github.com/Mudlet/Mudlet/pull/6102/commits/0fa9bfc6dc8c8a68a0954017baa7e4fea1195d8a .

vadi2 reviewed (commented):
Works great! Just have a couple of small look'n'feel suggestions.

SlySven committed:
Revise: remove a detail reckoned to be unnecessary from a tool tip

This was suggested by @vadi2.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: convert RoomID and RoomWeight widgets to be QLabels

However it is not possible to simply show an icon AND the room ID number
in the same `QLabel` in the manner I was previously using for a `QLineEdit`
(by adding a dummy `QAction`) so it was necessary to put a `QFrame` in the
place of the previous `QLineEdit` and stick `two `QLabels` (for the
room ID and for a padlock icon for a locked room) inside that with no
margins and arrange for the second to be hidden except when the room is
locked...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/ui/room_exits.ui:
```diff
@@ -757,16 +757,62 @@
+               <cursorShape>ForbiddenCursor</cursorShape>
+              </property>
+              <property name="toolTip">
+               <string>&lt;p&gt;This is the Room ID Number for this room - it cannot be changed here!</string>
```
```suggestion
               <string>&lt;p&gt;Room ID for this room</string>
```

Saying 'it cannot be changed here!' adds no value - you already know that because the field doesn't let you. It is important to keep tooltips as consise as possible.

SlySven committed:
Revise: remove phrase held to be redundant in a couple of tooltips.

This was suggested/requested by @vadi2.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>
